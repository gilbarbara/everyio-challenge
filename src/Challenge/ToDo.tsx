import { DragEvent, MouseEvent, useState } from 'react';
import { Box, Button, ButtonUnstyled, Flex, FlexInline, Icon } from '@gilbarbara/components';

import type { ToDo } from './config';
import { useAppState } from './utils/context';

export default function ToDo(props: ToDo) {
  const { columnId, content, id } = props;
  const [opacity, setOpacity] = useState(1);
  const {
    setAppState,
    state: { columns, todos },
  } = useAppState();

  const firstColumnId = columns[0].id;
  const lastColumnId = columns[columns.length - 1].id;

  const getTargetColumnId = (type: string) => {
    const index = columns.findIndex(column => column.id === columnId);
    const nextColumn = type === 'next' ? columns[index + 1] : columns[index - 1];

    if (!nextColumn) {
      return columnId;
    }

    return nextColumn.id;
  };

  const handleClickMove = (event: MouseEvent<HTMLButtonElement>) => {
    const { type = '' } = event.currentTarget.dataset;

    const canBeMoved = type === 'next' ? columnId !== lastColumnId : columnId !== firstColumnId;

    if (!columnId || !id || !canBeMoved) {
      return;
    }

    setAppState({
      todos: todos?.map(todo => {
        if (todo.id === id) {
          return {
            ...todo,
            columnId: getTargetColumnId(type),
          };
        }

        return todo;
      }),
    });
  };

  const handleClickRemove = (event: MouseEvent<HTMLButtonElement>) => {
    const { id: taskId = '' } = event.currentTarget.dataset;

    if (!taskId) {
      return;
    }

    setAppState({ todos: todos?.filter(todo => todo.id !== taskId) });
  };

  const handleDragStart = (event: DragEvent) => {
    event.dataTransfer.clearData();
    // Set the drag's format and data.
    // Use the event target's id for the data
    event.dataTransfer.setData('sourceId', event.currentTarget.id);
    setOpacity(0.6);
  };

  const handleDragEnd = () => {
    setOpacity(1);
  };

  return (
    <Box
      bg="white"
      data-column-id={columnId}
      data-component-name="ChallengeToDo"
      data-id={id}
      draggable
      id={id}
      onDragEnd={handleDragEnd}
      onDragStart={handleDragStart}
      opacity={opacity}
      p="xs"
      radius="md"
      shadow="low"
    >
      <Flex align="center" gap="xs" justify="space-between" textAlign="center">
        <Button
          bg="red.300"
          data-component-name="MoveToDoLeft"
          data-type="previous"
          disabled={columnId === firstColumnId}
          iconOnly
          onClick={handleClickMove}
          size="sm"
        >
          <Icon name="arrow-left" size={20} />
        </Button>
        <FlexInline data-component-name="ToDoContent" gap="xs">
          {content}
          <ButtonUnstyled
            aria-label="Remove task"
            data-component-name="RemoveTodo"
            data-id={id}
            onClick={handleClickRemove}
          >
            <Icon name="trash" />
          </ButtonUnstyled>
        </FlexInline>
        <Button
          bg="green.300"
          data-component-name="MoveToDoRight"
          data-type="next"
          disabled={columnId === lastColumnId}
          iconOnly
          onClick={handleClickMove}
          size="sm"
        >
          <Icon name="arrow-right" size={20} />
        </Button>
      </Flex>
    </Box>
  );
}
