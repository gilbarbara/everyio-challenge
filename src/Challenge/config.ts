export interface Column {
  id: string;
  name: string;
}

export interface ToDo {
  columnId: string;
  content: string;
  id: string;
}

export interface AppState {
  columns: Array<Column>;
  todos: Array<ToDo>;
}

export const appState = {
  columns: [
    { id: 'todo', name: 'To Do' },
    { id: 'in-progress', name: 'In Progress' },
    { id: 'done', name: 'Done' },
  ],
  todos: [] as Array<ToDo>,
} satisfies AppState;
