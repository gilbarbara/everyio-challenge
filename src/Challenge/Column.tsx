import { DragEvent } from 'react';
import { Flex, H4, NonIdealState } from '@gilbarbara/components';

import type { Column } from './config';
import ToDo from './ToDo';
import { useAppState } from './utils/context';

export default function Column(props: Column) {
  const { id, name } = props;
  const { setAppState, state } = useAppState();

  const columnTodos = state.todos.filter(todo => todo.columnId === id);

  const handleDragOver = (event: DragEvent) => {
    event.preventDefault();
  };

  const handleDrop = (event: DragEvent) => {
    const sourceId = event.dataTransfer.getData('sourceId');

    const todos = state.todos.map(todo => {
      if (todo.id === sourceId) {
        return {
          ...todo,
          columnId: id,
        };
      }

      return todo;
    });

    setAppState({ todos });
  };

  const content = columnTodos.length ? (
    columnTodos.map(todo => <ToDo key={todo.id} {...todo} />)
  ) : (
    <NonIdealState color="gray.300" description="No tasks yet" icon="tasklist" title={null} />
  );

  return (
    <Flex
      data-component-name="ChallengeColumn"
      data-id={id}
      direction="column"
      flex
      minHeight="60vh"
      p="md"
      radius="md"
      shadow="mid"
    >
      <H4 align="center">{name}</H4>
      <Flex
        align="stretch"
        data-component-name="TodoList"
        direction="column"
        flex
        gap="md"
        onDragOver={handleDragOver}
        onDrop={handleDrop}
      >
        {content}
      </Flex>
    </Flex>
  );
}
