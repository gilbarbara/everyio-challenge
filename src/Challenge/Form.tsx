import { FormEvent } from 'react';
import { Box, Button, Flex, Icon, Input } from '@gilbarbara/components';
import { createId } from '@paralleldrive/cuid2';

import { ToDo } from './config';
import { useAppState } from './utils/context';

export default function Form() {
  const {
    setAppState,
    state: { columns, todos },
  } = useAppState();

  const handleSubmitForm = (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    const formData = new FormData(event.currentTarget);
    const content = formData.get('content') as string;

    const newTodo: ToDo = {
      id: createId(),
      content,
      columnId: columns[0].id,
    };

    setAppState({
      todos: [...todos, newTodo],
    });

    event.currentTarget.reset();
  };

  return (
    <Box as="form" data-component-name="ChallengeForm" method="POST" onSubmit={handleSubmitForm}>
      <Flex gap="xs">
        <Input name="content" placeholder="Add a new task" required width="280px" />
        <Button type="submit">
          <Icon name="plus-heavy" />
        </Button>
      </Flex>
    </Box>
  );
}
