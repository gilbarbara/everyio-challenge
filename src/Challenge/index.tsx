import React from 'react';
import { Box, Flex, Paragraph } from '@gilbarbara/components';

import Column from './Column';
import Form from './Form';
import { StateProvider, useAppState } from './utils/context';

function ChallengeComponent() {
  const { state } = useAppState();

  return (
    <Box data-component-name="Challenge" p="lg">
      <Flex data-component-name="Columns" gap="xl">
        {state.columns.map(column => (
          <Column {...column} key={column.id} />
        ))}
      </Flex>
      <Flex justify="space-between" mt="lg">
        <Form />
        <Paragraph color="gray.400">You can also drag and drop between columns.</Paragraph>
      </Flex>
    </Box>
  );
}

export default function Challenge() {
  return (
    <StateProvider>
      <ChallengeComponent />
    </StateProvider>
  );
}
