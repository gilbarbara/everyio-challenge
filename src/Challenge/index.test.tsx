import { fireEvent, render, screen } from '@testing-library/react';

import Challenge from './index';

vi.mock('@gilbarbara/hooks', async () => {
  const hooks = await vi.importActual('@gilbarbara/hooks');

  return {
    ...hooks,
    // @ts-expect-error
    useLocalStorageState: (_: string, state: any) => hooks.useSetState(state),
  };
});

describe('Challenge', () => {
  it('should render properly', () => {
    render(<Challenge />);

    expect(screen.getByTestId('Challenge')).toMatchSnapshot();
  });

  it('should add a new task, move it around and remove it', () => {
    render(<Challenge />);

    fireEvent.change(screen.getByRole('textbox'), { target: { value: 'New task' } });
    fireEvent.click(screen.getByRole('button'));

    expect(screen.getByTestId('ChallengeToDo').dataset.columnId).toBe('todo');

    fireEvent.click(screen.getByTestId('MoveToDoRight'));
    expect(screen.getByTestId('ChallengeToDo').dataset.columnId).toBe('in-progress');

    fireEvent.click(screen.getByTestId('MoveToDoLeft'));
    expect(screen.getByTestId('ChallengeToDo').dataset.columnId).toBe('todo');

    fireEvent.click(screen.getByTestId('RemoveTodo'));
    expect(screen.queryByTestId('ChallengeToDo')).not.toBeInTheDocument();
  });
});
