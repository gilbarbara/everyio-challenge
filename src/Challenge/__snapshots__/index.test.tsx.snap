// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Challenge > should render properly 1`] = `
<div
  class="css-17djdg7"
  data-component-name="Challenge"
>
  <div
    class="css-wlnjiv"
    data-component-name="Columns"
  >
    <div
      class="css-16gbgy6"
      data-component-name="ChallengeColumn"
      data-id="todo"
    >
      <h4
        class="css-mgc95p"
        data-component-name="H4"
      >
        To Do
      </h4>
      <div
        class="css-1g9f7v4"
        data-component-name="TodoList"
      >
        <div
          class="css-ldpr0n"
          data-component-name="NonIdealState"
        >
          <div
            class="css-z5ifs4"
            data-component-name="Flex"
          >
            <span
              class="css-1l1reyq"
              data-component-name="Icon"
              name="tasklist"
            />
          </div>
          <div
            class="css-q34js2"
            data-component-name="Box"
          >
            <p
              class="css-1d6m1sj"
              data-component-name="Paragraph"
            >
              <span
                class="css-onif9v"
                data-component-name="Text"
              >
                No tasks yet
              </span>
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      class="css-16gbgy6"
      data-component-name="ChallengeColumn"
      data-id="in-progress"
    >
      <h4
        class="css-mgc95p"
        data-component-name="H4"
      >
        In Progress
      </h4>
      <div
        class="css-1g9f7v4"
        data-component-name="TodoList"
      >
        <div
          class="css-ldpr0n"
          data-component-name="NonIdealState"
        >
          <div
            class="css-z5ifs4"
            data-component-name="Flex"
          >
            <span
              class="css-1l1reyq"
              data-component-name="Icon"
              name="tasklist"
            />
          </div>
          <div
            class="css-q34js2"
            data-component-name="Box"
          >
            <p
              class="css-1d6m1sj"
              data-component-name="Paragraph"
            >
              <span
                class="css-onif9v"
                data-component-name="Text"
              >
                No tasks yet
              </span>
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      class="css-16gbgy6"
      data-component-name="ChallengeColumn"
      data-id="done"
    >
      <h4
        class="css-mgc95p"
        data-component-name="H4"
      >
        Done
      </h4>
      <div
        class="css-1g9f7v4"
        data-component-name="TodoList"
      >
        <div
          class="css-ldpr0n"
          data-component-name="NonIdealState"
        >
          <div
            class="css-z5ifs4"
            data-component-name="Flex"
          >
            <span
              class="css-1l1reyq"
              data-component-name="Icon"
              name="tasklist"
            />
          </div>
          <div
            class="css-q34js2"
            data-component-name="Box"
          >
            <p
              class="css-1d6m1sj"
              data-component-name="Paragraph"
            >
              <span
                class="css-onif9v"
                data-component-name="Text"
              >
                No tasks yet
              </span>
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="css-12xm4i9"
    data-component-name="Flex"
  >
    <form
      class="css-h2rdtk"
      data-component-name="ChallengeForm"
      method="POST"
    >
      <div
        class="css-c46fep"
        data-component-name="Flex"
      >
        <input
          class="css-dro6bi"
          data-component-name="Input"
          id="content"
          name="content"
          placeholder="Add a new task"
          required=""
        />
        <button
          class="css-7hon60"
          data-component-name="Button"
          type="submit"
        >
          <span
            class="css-1sm3hig"
            data-component-name="Icon"
            name="plus-heavy"
          />
        </button>
      </div>
    </form>
    <p
      class="css-v84yan"
      data-component-name="Paragraph"
    >
      <span
        class="css-3trm42"
        data-component-name="Text"
      >
        You can also drag and drop between columns.
      </span>
    </p>
  </div>
</div>
`;
